phygnn/layers/custom_layers.py

```diff
@@ -1344,20 +1423,119 @@ def call(x, hi_res_feature):
         hi_res_feature : tf.Tensor | np.ndarray
             This should be a 4D array for spatial enhancement model or 5D array
             for a spatiotemporal enhancement model (obs, spatial_1, spatial_2,
-            (temporal), 1) that can be used to fix values of x. This is NaN
-            where there are no observations and real values where observations
-            exist.
+            (temporal), 1). This is NaN where there are no observations and
+            real values where observations exist.
 
         Returns
         -------
         x : tf.Tensor
             Output tensor with the hi_res_feature used to fix values of x.
         """
+        if hi_res_feature is None:
+            hi_res_feature = tf.fill(x[..., :1].shape, np.nan)
         mask = tf.math.is_nan(hi_res_feature)
         fixed = tf.where(mask, x[..., :1], hi_res_feature)
         return tf.concat((x, fixed), axis=-1)
 
 
+class Sup3rConcatEmbeddedObs(tf.keras.layers.Layer):
+    """Layer to concatenate sparse data in the middle of a super resolution
+    forward pass, with a learned embedding. This is used to condition models
+    on sparse observation data."""
```

Comment:
I'm still not convinced that this is fundamentally different from letting the the generator learn how to handle the extra layers, and i'm confused about the implementation of the embed net attribute.

Reply:
This could make it easier to include the sparse data by encoding it instead of injecting physical values in latent space. I'm still seing pin points in the output and I think this will help. Also, this can encode obs within multiple channels instead of just one.

Reply:
Like i said, still not convinced :) but let me know what you find!